{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Excercise 1 (Math fun)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "__Given:__ Two positive integers a=100 and b=200\n",
    "\n",
    "__Return:__ The sum of all odd integers from a through b, inclusively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7500"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([_ for _ in range(101,200,2)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7500"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from functools import reduce\n",
    "reduce((lambda x, y: x + y), range(101,200,2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "__Given:__ If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.\n",
    "\n",
    "__Return:__ Find the sum of all the multiples of 3 or 5 below 1000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "233168"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([_ for _ in range(0,1000) if (not(_%3) or not(_%5))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "233168"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([i for i in range(1000) if i%3==0 or i%5==0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "__Given:__ A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 Ã— 99.\n",
    "\n",
    "__Return:__ Find the largest palindrome made from the product of two 3-digit numbers.\n",
    "\n",
    "__Hint:__ To learn how to check for palindromism try out \"lagerregal\"[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(906609, 913, 993)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "collect = []\n",
    "for i in range(100,999):\n",
    "    for j in range(i,1000):\n",
    "        prod = str(i*j)\n",
    "        if prod==prod[::-1]:\n",
    "            collect.append((i*j,i,j))\n",
    "sorted(collect, key = lambda x: x[0], reverse = True)[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "__Given:__ Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be: 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...\n",
    "\n",
    "__Return:__ By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.\n",
    "\n",
    "__Hint:__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "def fib():\n",
    "    a, b = 1, 2\n",
    "    while 1:\n",
    "        yield(a)\n",
    "        a, b = b, a + b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4613732"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fib_sum = 0\n",
    "for x in fib():\n",
    "    if x>4000000:\n",
    "        break\n",
    "    if not(x%2):\n",
    "        fib_sum+=x\n",
    "fib_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4613732"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fib_sum = 0\n",
    "count = 0\n",
    "for f in fib():\n",
    "    if f>4*10**6:\n",
    "        break\n",
    "    elif f%2==0:\n",
    "        fib_sum += f\n",
    "fib_sum"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Excercise 2 (string/bioinformatics fun)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "transcribe the DNA into RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "s='GATGGAACTTGACTACGTAAATT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GAUGGAACUUGACUACGUAAAUU'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s.replace('T','U')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "report the nucleotide counts for each sequence individually\n",
    "\n",
    "__hint__: split at '>'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = '''>sequence 1\n",
    "CCTGCGGAAGATCGGCACTAGAATAGCCAGAACCGTTTCTCTGAGGCTTCCGGCCTTCCC\n",
    "TCCCACTAATAATTCTGAGG\n",
    ">sequence 2\n",
    "CCATCGGTAGCGCATCCTTAGTCCAATTAAGTCCCTATCCAGGCGCTCCGCCGAAGGTCT\n",
    "ATATCCATTTGTCAGCAGACACGC\n",
    ">sequence 3\n",
    "CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGAC\n",
    "TGGGAACCTGCGGGCAGTAGGTGGAAT\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence 1 Counter({'C': 25, 'T': 19, 'G': 18, 'A': 18})\n",
      "sequence 2 Counter({'C': 28, 'T': 20, 'A': 19, 'G': 17})\n",
      "sequence 3 Counter({'G': 29, 'C': 24, 'A': 20, 'T': 14})\n"
     ]
    }
   ],
   "source": [
    "for seq in s.split('>')[1:]:\n",
    "    seq = seq.split('\\n')\n",
    "    print(seq[0], Counter(list(''.join(seq[1:]))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "convert the list of gene symbols into entrez and ensembl identifier\n",
    "\n",
    "__hint 1__ split at newline\n",
    "\n",
    "__hint 2__ use use mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "symbols = '''BRD4\n",
    "CDK2\n",
    "KRAS\n",
    "TP53\n",
    "MYC\n",
    "CTNNB1\n",
    "ATAD2\n",
    "PLK1\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "symbols = symbols.split('\\n')\n",
    "symbols.remove('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "mg = mygene.MyGeneInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-8...done.\n",
      "Finished.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>ensembl.gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>BRD4</td>\n",
       "      <td>23476</td>\n",
       "      <td>90.992455</td>\n",
       "      <td>23476</td>\n",
       "      <td>ENSG00000141867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CDK2</td>\n",
       "      <td>1017</td>\n",
       "      <td>86.533280</td>\n",
       "      <td>1017</td>\n",
       "      <td>ENSG00000123374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>88.561424</td>\n",
       "      <td>3845</td>\n",
       "      <td>ENSG00000133703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TP53</td>\n",
       "      <td>7157</td>\n",
       "      <td>87.215040</td>\n",
       "      <td>7157</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MYC</td>\n",
       "      <td>4609</td>\n",
       "      <td>86.693140</td>\n",
       "      <td>4609</td>\n",
       "      <td>ENSG00000136997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>1499</td>\n",
       "      <td>84.942560</td>\n",
       "      <td>1499</td>\n",
       "      <td>ENSG00000168036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ATAD2</td>\n",
       "      <td>29028</td>\n",
       "      <td>87.987816</td>\n",
       "      <td>29028</td>\n",
       "      <td>ENSG00000156802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PLK1</td>\n",
       "      <td>5347</td>\n",
       "      <td>85.326960</td>\n",
       "      <td>5347</td>\n",
       "      <td>ENSG00000166851</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          _id     _score entrezgene     ensembl.gene\n",
       "query                                               \n",
       "BRD4    23476  90.992455      23476  ENSG00000141867\n",
       "CDK2     1017  86.533280       1017  ENSG00000123374\n",
       "KRAS     3845  88.561424       3845  ENSG00000133703\n",
       "TP53     7157  87.215040       7157  ENSG00000141510\n",
       "MYC      4609  86.693140       4609  ENSG00000136997\n",
       "CTNNB1   1499  84.942560       1499  ENSG00000168036\n",
       "ATAD2   29028  87.987816      29028  ENSG00000156802\n",
       "PLK1     5347  85.326960       5347  ENSG00000166851"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg.querymany(symbols, scopes='symbol', species=9606, fields = ['entrezgene','ensembl.gene'], as_dataframe = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "translate RNA into protein\n",
    "AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\n",
    "\n",
    "__Hint:__ use the RNA codon [table](http://rosalind.info/glossary/rna-codon-table/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA = 'AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "trans = '''UUU F      CUU L      AUU I      GUU V\n",
    "UUC F      CUC L      AUC I      GUC V\n",
    "UUA L      CUA L      AUA I      GUA V\n",
    "UUG L      CUG L      AUG M      GUG V\n",
    "UCU S      CCU P      ACU T      GCU A\n",
    "UCC S      CCC P      ACC T      GCC A\n",
    "UCA S      CCA P      ACA T      GCA A\n",
    "UCG S      CCG P      ACG T      GCG A\n",
    "UAU Y      CAU H      AAU N      GAU D\n",
    "UAC Y      CAC H      AAC N      GAC D\n",
    "UAA Stop   CAA Q      AAA K      GAA E\n",
    "UAG Stop   CAG Q      AAG K      GAG E\n",
    "UGU C      CGU R      AGU S      GGU G\n",
    "UGC C      CGC R      AGC S      GGC G\n",
    "UGA Stop   CGA R      AGA R      GGA G\n",
    "UGG W      CGG R      AGG R      GGG G'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = np.concatenate([list(filter(None,ele.split())) for ele in trans.split('\\n')])\n",
    "trans_d = {}\n",
    "for v, w in zip(tmp[:-1:2], tmp[1::2]):\n",
    "    trans_d[v]=w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MAMAPRTEINSTRING'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''.join([trans_d[triplet] for triplet in re.findall('...',RNA) if trans_d[triplet] != 'Stop'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Excercise 3 (Data science fun)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "## Problem 1\n",
    "What is the pearson correlation of MDM2 copy number and it's expression in CCLE cell lines? Also generate a scatter plot!\n",
    "\n",
    "__hints__:\n",
    "* download CN data from [depmap](https://ndownloader.figshare.com/files/16757699)\n",
    "* join MDM2 copynumber data to expression\n",
    "* find the pearson correlation function in scipy stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "In which tumor/tissue type within the CCLE dataset is the highest expression median expression of KLK3?\n",
    "\n",
    "__hints__:\n",
    "* join sample_info to expression\n",
    "* use groupby apply\n",
    "* generate a boxplot with the tissuetype as categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "Generate a PCA plot using the top100 most varying genes (regarding expression)! Color cell lines according to tissue type\n",
    "\n",
    "__hints__:\n",
    "* read and use https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "Generate a umap projection on same top100 most varying genes (also with tissue type as colors)\n",
    "\n",
    "__hints__:\n",
    "* read and use https://umap-learn.readthedocs.io/en/latest\n",
    "* use euclidean distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python_workshop_2019",
   "language": "python",
   "name": "python_workshop_2019"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
